File: shcloj2/code/src/examples/midi.clj
;;
;;Excerpted from "Programming Clojure, Second Edition",
;;published by The Pragmatic Bookshelf.
;;Copyrights apply to this code. It may not be used to create training material, 
;;courses, books, articles, and the like. Contact us if you are in doubt.
;;We make no guarantees that this code is fit for any purpose. 
;;Visit http://www.pragmaticprogrammer.com/titles/shcloj2 for more book information.
;;
(ns examples.datatypes.midi
  (:import [javax.sound.midi MidiSystem]))
(defprotocol MidiNote
  (to-msec [this tempo])
  (key-number [this])
  (play [this tempo midi-channel]))

(defn perform [notes & {:keys [tempo] :or {tempo 88}}]
  (with-open [synth (doto (MidiSystem/getSynthesizer).open)]
    (let [channel (aget (.getChannels synth) 0)]
      (doseq [note notes]
        (play note tempo channel)))))

(defrecord Note [pitch octave duration]
  MidiNote
  (to-msec [this tempo]
    (let [duration-to-bpm {1 240, 1/2 120, 1/4 60, 1/8 30, 1/16 15}]
      (* 1000 (/ (duration-to-bpm (:duration this))
                 tempo))))
  (key-number [this]
    (let [scale {:C 0,  :C# 1, :Db 1,  :D 2,
                 :D# 3, :Eb 3, :E 4,   :F 5,
                 :F# 6, :Gb 6, :G 7,   :G# 8,
                 :Ab 8, :A 9,  :A# 10, :Bb 10,
                 :B 11}]
      (+ (* 12 (inc (:octave this)))
         (scale (:pitch this)))))
  (play [this tempo midi-channel]
    (let [velocity (or (:velocity this) 64)]
      (.noteOn midi-channel (key-number this) velocity)
      (Thread/sleep (to-msec this tempo)))))

File: elixir16/code/protocols/midi.exs
#---
# Excerpted from "Programming Elixir ≥ 1.6",
# published by The Pragmatic Bookshelf.
# Copyrights apply to this code. It may not be used to create training material,
# courses, books, articles, and the like. Contact us if you are in doubt.
# We make no guarantees that this code is fit for any purpose.
# Visit http://www.pragmaticprogrammer.com/titles/elixir16 for more book information.
#---
defmodule Midi do

  defstruct(content: <<>>)
  
  defmodule Frame do
    defstruct(
      type:   "xxxx",
      length: 0,
      data:   <<>>
    )

    def to_binary(%Midi.Frame{type: type, length: length, data: data}) do
      <<
           type::binary-4,
         length::integer-32,
           data::binary
      >>
    end
  end

  def from_file(name) do
    %Midi{content: File.read!(name)}
  end
end


defimpl Enumerable, for: Midi do

  def reduce(%Midi{content: content}, state, fun) do
    _reduce(content, state, fun)
  end  
  
  def _reduce(_content, {:halt, acc}, _fun) do
    {:halted, acc}
  end
  
  def _reduce(content, {:suspend, acc}, fun) do
    {:suspended, acc, &_reduce(content, &1, fun)}
  end

  def _reduce(_content = "", {:cont, acc}, _fun) do
    {:done, acc}
  end
  
  def _reduce(<<
                  type::binary-4,
                length::integer-32,
                  data::binary-size(length),
                  rest::binary
              >>,
              {:cont, acc},
              fun
             ) do
    frame = %Midi.Frame{type: type, length: length, data: data}
    _reduce(rest, fun.(frame, acc), fun)
  end

  def count(midi = %Midi{}) do
    frame_count = Enum.reduce(midi, 0, fn (_, count) -> count+1 end)
    { :ok, frame_count }
  end

  def member?(%Midi{}, %Midi.Frame{}) do
    { :error, __MODULE__ }
  end

  def slice(%Midi{}) do
    { :error, __MODULE__ }
  end
  
end


defimpl Collectable, for: Midi do
  use Bitwise

  def into(%Midi{content: content}) do
    {
      content,
      fn
        acc, {:cont, frame = %Midi.Frame{}} ->
          acc <> Midi.Frame.to_binary(frame)
        
        acc,  :done ->
          %Midi{content: acc}
        
        _, :halt ->
          :ok
      end
    }
  end
end


